### ISSUE TITLE
Query Comments

### ISSUE THREAD

**Issue by Ustice** [prisma/prisma#10746]
## Problem

I need a way to trace queries to the request that spawned them, but unfortunately Prisma doesn't offer any lifecycle hooks that I can use. After looking for some workarounds, I came across [SQL Commenter](https://google.github.io/sqlcommenter/), which seems to offer a very simple workaround, but unfortunately, there does not currently seem to be a way to add comments to a query.

## Suggested solution

It should be easy enough to add SQL comments by adding a `comment` method to the prisma client. For instance:

```JavaScript
const comment = `Request Id: ${ requestId }`
const user = await prisma.comment().user.findUnique({
  where: {
    email: 'elsa@prisma.io',
  },
})
```

```SQL
/* Request Id: ff849408-5170-45f7-a4c9-85c844ff61f0 */
SELECT * FROM user WHERE email = "elsa@prisma.io";
```

## Alternatives

An alternative solution, and one that I was originally looking for was to be able to add metadata to queries. These could be using in the event-based logging to identify the source of query. Unfortunately, this doesn't exist either.

This could be expanded to allow for SQL block comments and the like, but even just a query-level comment would be _very_ powerful.

**Comment by sergioengineer** [comment]
I was also looking for this.

Do you feel like being able to return "request id" directly with the request would be more or as helpful? Like so:
`select "ff849408-5170-45f7-a4c9-85c844ff61f0" as request_id, .... FROM ....`


Because then we could maybe come up with a solution that would solve your issue and [this one](https://github.com/prisma/prisma/issues/4433) here which is about changing field names.

I'd be down to send PR for this. In one week or so I'll have some spare time.

**Comment by Ustice** [comment]
That certainly could be useful, but it might be more ambiguous. That said, if that feature became available I would use it for this purpose.

**Comment by KhanhHH** [comment]
This one sounds good, really helpful feature!

**Comment by Ustice** [comment]
@sergioengineer I've learned more about how Prisma works under the hood since when I wrote the original suggestion. Did you try work on that PR? I'm still figuring out how the client works.

**Comment by adamatti** [comment]
+1 here, I am coming from typeorm and missing this feature

**Comment by MatheusLasserre** [comment]
I would love this one. Most of the time i use PlanetScale as my db, so comments would be really useful to track error and query times, since they support tagging query by comments.

And to be honest, that's kind of a big deal when the project scales and performance starts to become more and more important

**Comment by janpio** [comment]
Related: https://github.com/prisma/prisma/issues/13360

**Comment by eliliam** [comment]
Any updates here? This is a crucial feature with a large codebase, currently it is very difficult to track down where queries are coming from, especially in cases where multiple places are running similar/identical queries making it impossible to distinguish from query logs alone

**Comment by StanleySathler** [comment]
Not to mention it'd be useful to track these queries on GCP Query Insights too.

**Comment by Aarbel** [comment]
Yes that's very useful, otherwise sql logs are just very opaque (for a large software)

**Comment by janpio** [comment]
Please @ mention team members or other users if you do not actually respond to what they have written something specifically for them. This is inappropriate.

**Comment by Jolg42** [comment]
Did anyone try the `tracing` preview feature?

See https://www.prisma.io/blog/tracing-tutorial-prisma-pmkddgq1lm2
Docs: https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/opentelemetry-tracing

The SQL queries will automatically be annotated with a comment like
```
/* traceparent='00-71a4cbd5fc3c2cf97b4cbbeddea28cb3-6a2c496ff1211ea4-01' */
```

Example from the blog post:
```
app.get("/users/random", async (_req: Request, res: Response) => {

     await tracer.startActiveSpan("GET /users/random", async (requestSpan) => {
        try {
            let users = await prisma.user.findMany({
                include: {
                    posts: true
                }
            });

            // select 10 users randomly
            const shuffledUsers = users.sort(() => 0.5 - Math.random());
            const selectedUsers = shuffledUsers.slice(0, 10);

            requestSpan.setAttribute("http.status", 200);
            res.status(200).json(selectedUsers);
        } catch (e) {

            requestSpan.setAttribute("http.status", 500);
            res.status(500).json({ error: 500, details: e });

        } finally {

             requestSpan.end();
        }
    });
});
```
I guess that could help, depending on what you want to do.

**Comment by Aarbel** [comment]
Thanks @Jolg42 .

That could surely help, that's still a little bit heavy to implement. The best would be to be able to simply add a comment in the SQL query generated by prisma.

Examples

```JS
let users = await prisma.user.findMany({
    comment: "GET /users/random",
    include: {
        posts: true
    }
});
```

or 

```JS
const user = await prisma.comment("GET /users/random").user.findUnique({
  where: {
    email: 'elsa@prisma.io',
  },
})
```

or any kind of other idea that helps to add custom comments inside queries for observability

**Comment by brian-dlee** [comment]
I agree with the need for this. I can see how tracing can certainly help with identification. My issue is that slow queries show up in the logs, but tracking down which Prisma query in the application actually sourced the query is very difficult to track down. Even if I had tracing, I would have to look at the slow query log and then cross check the application logs to find what kind of behaviors led to it.

I tend to agree with other proposals that a simple comment addition would provide a means of finding things. If it's possible then autogenerating a comment based on the source code file and line number in which the query was generated from would be excellent.

I do intend to look into the tracing integrations further, but I wanted to add my voice and use case to the comment feature request.

For what it's worth, I did look at the tracing several months ago. I had trouble seeing how I was able to integrate telemetry built-in to Google Cloud Logging. For the case of Google Cloud Logging to work, I need to be able to log (with my application logger) a JSON string that includes special fields such as `"logging.googleapis.com/trace": "projects/{project}/traces/{traceIdFromReq}"` and `"logging.googleapis.com/spanId": "{spanIdFromReq}"`. I don't know if things have changed, but the logging function available to me via the Prisma client were not capably of passing information from my request through to the logging handlers so I could achieve this. It kind of stopped me in my tracks.

**Comment by ogroppo** [comment]
Any update? Very requested feature over here

**Comment by dizzycoder1112** [comment]
I use a tricky way via `EventEmitter` to implement it

```
// traceMiddleware.ts

import { v4 as uuidv4 } from 'uuid'
import { asyncContext, queryEventEmitter } from "@monorepo-packages/db";


export function traceMiddleware(req: any, res: any, next: any) {
  const traceId = uuidv4()
    res.setHeader('X-Trace-Id', traceId)
    queryEventEmitter.emit('traceId', traceId);
    asyncContext.run({ traceId }, () => {
      queryEventEmitter.once('query:start', (event) => {
        event = JSON.parse(event);
        console.log(event)
        console.log("traceId", traceId)
      });
      next(); 
    });
}

```

```
// db.ts

export const asyncContext = new AsyncLocalStorage<TraceContext>();
export const queryEventEmitter = new EventEmitter();

export function attachLogging(prisma: PrismaLike, serviceName = 'unknown') {
  prisma.$on('query', (e) => {
    queryEventEmitter.emit('query:start', JSON.stringify(e));
  });
}
```

When the API is triggered, the terminal would show the details like

```
{
  timestamp: '2025-03-29T20:59:00.346Z',
  query: 'SELECT "public"."users"."id", "public"."users"."nickname", "public"."users"."created_at", "public"."users"."updated_at" FROM "public"."users" WHERE 1=1 OFFSET $1',
  params: '[0]',
  duration: 3,
  target: 'quaint::connector::metrics'
}
traceId 6a029450-ef03-4e0e-9800-ffbc078c6c86

```
I hope this way will be helpful for everyone who is suffering from this issue.

**Comment by rinslow** [comment]
@timsuchanek can we get attention to this issue? It's been long requested and a must for heavy production systems

**Comment by aqrln** [comment]
@rinslow please refrain from spamming and abusing the mention functionality, your behaviour violates [GitHub's Aceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-disrupting-the-experience-of-other-users) and can lead to consequences if it is repeated. The person you tagged has not written anything in this issue and you are not responding to them, they also cannot help you with your request.

You can see our current roadmap at https://github.com/prisma/prisma/issues/27503. Although unfortunately this quarter we haven't been able to allocate as much time as we planned for the community issues like we did last quarter, they are still top of our mind, and we'll continue to prioritize them based on the upvotes and other criteria.

As for this issue, yes, observability is indeed a must for production systems. That's why we implemented built-in OpenTelemetry support, which, among other things, adds the `traceparent` comments to the generated SQL queries. It has been in general availability since version 6.1.0 and doesn't require using preview features. It is an industry standard, and we know about large scale and high load projects using Prisma that rely on OpenTelemetry to solve the problems raised in this issue and are quite happy with it.

We do, however, recognize that it doesn't cover all possible use cases, and there may still be need to add more metadata to the queries than just being able to correlate them with the application code spatially and temporally (both of which OpenTelemetry solves: you can see the full timeline and hierarchy of calls, and you can correlate the queries with exact locations in code by instrumenting your application with spans). That's why this issue is still open.

**Comment by moltar** [comment]
> which, among other things, adds the traceparent comments to the generated SQL queries

Hey do you have a link for that? Does it allow us to inject extra stuff in there? Thanks.

**Comment by aqrln** [comment]
Prisma 7.1.0 added sqlcommenter support: https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/sql-comments. For things like request IDs, consider using the [`@prisma/sqlcommenter-query-tags`](https://www.npmjs.com/package/@prisma/sqlcommenter-query-tags) plugin.
